### PR TITLE
Parallelization via OpenMP support

### PR DESCRIPTION
**Pull request by Abde5**
Dear Wei, Liu, Ling and Su,

Reading the Discussion chapter of your paper, we took the freedom to implement a parallelization of CoACD. 

We did so by parallelizing part of the Compute problem (each CoACD iteration is perfectly parallelizable) and the merge (the cost matrix computing has been parallelized). We attempted to implement a more complex parallelization system with OpenMP tasks; however, it did not give good results compared to a basic parallelization of the for routines.

The Merge loop has been simplified in order to keep it a single loop. 

With empirical tests, we have observed that we achieve speedups from 1.2 to 3.8 for the smaller concavity thresholds.

chrome_yjWLeCmtUX

Please be free to try it out and I would be glad to answer any questions about the implementation.

This work is being contributed by [Altheria Solutions](https://altheria.com/).

### CODE REVIEW

File: src/process.cpp [hunk 1]
```diff
@@ -18,248 +20,275 @@ void ManifoldPreprocess(Params &params, Model &m, ofstream &of)
...
-          costMatrix[idx] = ComputeHCost(cvxs[p1], cvxs[p2], combinedCH, params.rv_k, params.resolution, params.seed);
-          precostMatrix[idx++] = max(ComputeHCost(meshs[p1], cvxs[p1], params.rv_k, 3000, params.seed),
-                                     ComputeHCost(meshs[p2], cvxs[p2], params.rv_k, 3000, params.seed));
-        }
-        else
-        {
-          costMatrix[idx++] = INF;
+        size_t p1, p2;
+#ifdef _OPENMP
+#pragma omp parallel for default(none) shared(costMatrix,precostMatrix,cvxs,params) private(p1,p2)
```
Comment (olitheolix) [inline src/process.cpp:51]:
```suggestion
#pragma omp parallel for default(none) shared(costMatrix,precostMatrix,cvxs,params,bound,threshold,meshs) private(p1,p2)
```
Make it work on Linux.

File: src/process.cpp [hunk 2]
```diff
@@ -18,248 +20,275 @@ void ManifoldPreprocess(Params &params, Model &m, ofstream &of)
...
-      Plane bestplane;
-      pmesh.ComputeVCH(pCH);
-      double h = ComputeHCost(pmesh, pCH, params.rv_k, params.resolution, params.seed, 0.0001, false);
+        vector<Model> tmp;
+        of << "iter " << iter << " ---- "
+            << "waiting pool: " << InputParts.size() << endl;
+        cout << "iter " << iter << " ---- "
+            << "waiting pool: " << InputParts.size() << endl;
+#ifdef _OPENMP
+#pragma omp parallel for default(none) shared(InputParts,params) private(cut_area)
```
Comment (olitheolix) [inline src/process.cpp:201]:
```suggestion
#pragma omp parallel for default(none) shared(InputParts,params,mesh,writelock,parts,pmeshs,tmp,cout) private(cut_area)
```
Make it work on Linux.

**Review by olitheolix — COMMENTED** [review]
@Abde5 It did not compile on Linux out of the box due to various
```
.../process.cpp:229:37: error: ‘pmeshs’ not specified in enclosing ‘parallel’
```
errors. However, I managed to make them disappear with the two suggested changes. CoACD now uses all cores during the decomposition phase of the algorithm.

Thank you for that PR :rocket:

**Comment by Abde5** [comment]
@olitheolix to be honest, the PR did not even compile on Windows 😅 

I didn't PR with the latest commit in my local branch, thanks for fixing the problem, tho!

**Comment by SarahWeiii** [comment]
Hi @Abde5 ,
Thank you for the contribution, I have tested the parallelization on Ubuntu 20.04 and will merge it into the main branch.

**Pull request merged**